File: go/v1/statement.go
/*
Wrapper APIs for in-toto attestation Statement layer protos.
*/

package v1

import "errors"

const StatementTypeUri = "https://in-toto.io/Statement/v1"

func (s *Statement) Validate() error {
	if s.GetType() != StatementTypeUri {
		return errors.New("Wrong statement type")
	}

	if s.GetSubject() == nil || len(s.GetSubject()) == 0 {
		return errors.New("At least one subject required")
	}

	// check all resource descriptors in the subject
	subject := s.GetSubject()
	for _, rd := range subject {
		if err := rd.Validate(); err != nil {
			return err
		}

		// v1 statements require the digest to be set in the subject
		if len(rd.GetDigest()) == 0 {
			return errors.New("At least one digest required")
		}
	}

	if s.GetPredicateType() == "" {
		return errors.New("Predicate type required")
	}

	if s.GetPredicate() == nil {
		return errors.New("Predicate object required")
	}

	return nil
}

File: python/in_toto_attestation/v1/statement.py
#  Wrapper class for in-toto attestation Statement protos.

from in_toto_attestation._internal.in_toto_attestation.v1 import Statement as _Statement
from in_toto_attestation.v1.resource_descriptor import ResourceDescriptor

# import in_toto_attestation.v1.statement_pb2 as spb
# from in_toto_attestation.v1.resource_descriptor import ResourceDescriptor

STATEMENT_TYPE_URI = "https://in-toto.io/Statement/v1"


class Statement(_Statement):
    def validate(self) -> None:
        if self.type != STATEMENT_TYPE_URI:
            raise ValueError("Wrong statement type")

        if len(self.subject) == 0:
            raise ValueError("At least one subject required")

        # check all resource descriptors in the subject
        subject = self.subject
        for i, rdpb in enumerate(subject):
            rd = ResourceDescriptor().from_dict(rdpb.to_dict())
            rd.validate()

            # v1 statements require the digest to be set in the subject
            if len(rd.digest) == 0:
                # return index in the subjects list in case of failure:
                # can't assume any other fields in subject are set
                raise ValueError(f"At least one digest required (subject {i})")

        if self.predicate_type == "":
            raise ValueError("Predicate type required")

        if len(self.predicate) == 0:
            raise ValueError("Predicate object required")
